src/core/mac/mac.cpp

```diff
@@ -650,23 +628,12 @@ otError Mac::SetExtendedPanId(const otExtendedPanId &aExtendedPanId)
     return OT_ERROR_NONE;
 }
 
-otError Mac::SendFrameRequest(Sender &aSender)
+otError Mac::SendFrameRequest(void)
 {
     otError error = OT_ERROR_NONE;
 
     VerifyOrExit(mEnabled, error = OT_ERROR_INVALID_STATE);
-    VerifyOrExit(mSendTail != &aSender && aSender.mNext == NULL, error = OT_ERROR_ALREADY);
-
-    if (mSendHead == NULL)
-    {
-        mSendHead = &aSender;
-        mSendTail = &aSender;
-    }
-    else
-    {
-        mSendTail->mNext = &aSender;
-        mSendTail        = &aSender;
-    }
+    VerifyOrExit(!mPendingTransmitData, error = OT_ERROR_ALREADY);
```

Comment:
`mPendingTransmitData` is cleared in `Mac::PerformOperation()`. Should we return error here even when a transmission is in progress?

Reply:
In the first implementation I actually had this:

```VerifyOrExit(!mPendingTrasnmitData && (mOperation != kOperationTransmitData), ...);``` 

but decided to remove it. 

- In `Mac::Mac::HandleTransmitDone()` we do call `HandleFrameSent()` on mesh-forwarder and then do `FinishOperation()` (so while in `HandleFrameSent()` callback the `mOperation` is still `kOperationTransmitData`).
- I think it's good to allow the next layer to request a new transmission from `HandleFrameSent()` call (even though current `MeshForwarder` won't do this and will post a tasklet).  In `Sender` model we allowed this by removing the `Sender` from link-list before invoking the callback, so caller could request/register same `Sender`.
- With the `mPendingTransmitData` model we can technically schedule a pending operation even if there is an ongoing tx operation.

What do you think?

Reply:
I agree that when the callback to the higher layer indicates that the transmission is complete, the higher layer should be able to submit a new request.

I'm not clear on the benefit of requesting another transmission before the callback to the higher layer indicates that the current transmission is complete. Other operations like ActiveScan and EnergyScan don't allow another request until the callback to the higher layer indicates those operations are complete.